File: Tuning/sample_code/03_prof-ex/src/fortran/calc_iniposit.f90
! Copyright 2024 Research Organization for Information Science and Technology
!!$---------------------------------------------------------------------
!!$ calc_iniposit.f90
!!$ Author: Yukihiro Ota (yota@rist.or.jp)
!!$ Last update: 1 Nov 2016
!!$---------------------------------------------------------------------

subroutine calc_iniposit(icon,NRD)

  use common_par,only: num_partot,unit_std_out,leng
  
  use common_var,only: rx0,ry0,rz0

  implicit none
  integer,intent(inout) :: icon
  integer,intent(in) :: NRD

  !! /* local variables */
  integer :: ii,ij
  integer :: ix,iy,iz
  integer :: iok,ird,ioffset,icount
  integer,parameter :: MAXCOUNT=100

  double precision :: crx0,cry0,crz0
  double precision :: rxij,ryij,rzij,rsqij
  double precision :: rd(1:NRD)

  !! generate a random sequence
  call get_rnd_array(NRD,rd)
  
  !! get position of 1st atom 
  rx0(1) = leng*( rd(1)-0.5d0 )
  ry0(1) = leng*( rd(2)-0.5d0 )
  rz0(1) = leng*( rd(3)-0.5d0 )

  !! generate position of other atoms
  iok     = 0
  ii      = 2
  ird     = 3
  icount  = 1
  ioffset = 2
  loopii:do while ( ii .le. num_partot )

     ix = 3*(ioffset-1) + 1
     iy = ix + 1
     iz = ix + 2
     ird = ird + 3
     
     crx0 = leng*( rd(ix)-0.5d0 )
     cry0 = leng*( rd(iy)-0.5d0 )
     crz0 = leng*( rd(iz)-0.5d0 )

     !! check inter-atom distance
     do ij = 1, ii-1

        rxij = crx0 - rx0(ij)
        ryij = cry0 - ry0(ij)
        rzij = crz0 - rz0(ij)

        !! periodic boundary condition
        rxij = rxij - dnint( rxij/leng )*leng
        ryij = ryij - dnint( ryij/leng )*leng
        rzij = rzij - dnint( rzij/leng )*leng
        
        !! distance criterion: removing too short
        !! inter-atom distance
        rsqij = rxij*rxij + ryij*ryij + rzij*rzij
        if ( rsqij .lt. 1.0d0 ) then
           ioffset = ioffset + 1
           iok = 1
           exit
        end if
        
     end do

     !! update if iok=0 (and otherwise ... )
     if ( iok .eq. 0 ) then
        rx0(ii) = crx0
        ry0(ii) = cry0
        rz0(ii) = crz0
        ii = ii + 1
     else
        !! if no remaining random variable
        if ( ird .ge. NRD-3 ) then
           call get_rnd_array(NRD,rd)
           ioffset = 1
           ird = 0
           icount = icount + 1
        end if
        iok = 0
     end if

     !! check MAXCOUNT
     if ( icount .eq. MAXCOUNT ) then
        icon = -1 
        exit
     end if
     
  end do loopii

  !! Finalization
  if ( icon .lt. 0 ) then
     write(unit_std_out,'(A)') &
          "Initialization of particle position is failed."
     write(unit_std_out,'(A)') &
          "This error may often consider a too-dense case."
     write(unit_std_out,'(A)') &
          "You carefully tune total number of particles and "
     write(unit_std_out,'(A)') &
          "unit-cell size."
     return
  end if
  
  icon = 0
  return
end subroutine calc_iniposit

File: Tuning/sample_code/03_prof-ex/src/c/calc_iniposit.c
/* Copyright 2024 Research Organization for Information Science and Technology */
/*-------------------------------------------------------------------*/
/* calc_iniposit.c                                                            
   Author: Yukihiro Ota (yota@rist.or.jp)                            
   Last update: 31 July 2017                                         */
/*-------------------------------------------------------------------*/
#include <stdio.h>
#include <math.h>
#include "setup.h"
#include "rndvar.h"
#include "calc.h"

#ifdef __DEBUG__
#include <stdlib.h>

int calc_iniposit (int nn, InputData *IDp, AtomData *ADp)
{
  int ierr=0;
  char cbuff[2];
  double x,y,z;

  FILE *fp;
  char *filename="geometry.dat";
  fp = fopen(filename,"r");
  for ( int i=0; i < IDp->num_partot; i=i+1 ) {
    fscanf(fp,"%s %lE %lE %lE",cbuff,&x,&y,&z);
    ADp->rx0[i] = x;
    ADp->ry0[i] = y;
    ADp->rz0[i] = z;
  }
  fclose(fp);
  return ierr;
}
#else
int calc_iniposit (int nn, InputData *IDp, AtomData *ADp)
{
  int ierr=0;
  int ii,ij;
  int ix,iy,iz;
  int iok,ird,ioffset,icount;
  const int MAXCOUNT=100;

  double crx0,cry0,crz0;
  double rxij,ryij,rzij,rsqij;
  double rd[nn]; 

  /* get cell length */
  double leng = IDp->leng;

  /* generate random sequence */
  get_rnd_array(&nn,rd);

  /* get position of 1st atom */
  ADp->rx0[0] =  leng * ( rd[0]-0.5 );
  ADp->ry0[0] =  leng * ( rd[1]-0.5 );
  ADp->rz0[0] =  leng * ( rd[2]-0.5 );

  /* generate position of other atoms */
  iok     = 0;
  ii      = 1;
  ird     = 3;
  icount  = 1;
  ioffset = 2;
  while ( ii < IDp->num_partot ) {

    ix  = 3*(ioffset-1);
    iy  = ix  + 1;
    iz  = ix  + 2;
    ird = ird + 3;

    crx0 = leng * ( rd[ix]-0.5 );
    cry0 = leng * ( rd[iy]-0.5 );
    crz0 = leng * ( rd[iz]-0.5 );

    /* check inter-atom distance */
    for ( ij=0; ij < ii; ij=ij+1) {

      rxij = crx0 - ADp->rx0[ij]; 
      ryij = cry0 - ADp->ry0[ij]; 
      rzij = crz0 - ADp->rz0[ij];

      /* periodic boundary condition (with round in c99) */ 
      rxij = rxij - round( rxij/leng )*leng;
      ryij = ryij - round( ryij/leng )*leng;
      rzij = rzij - round( rzij/leng )*leng;

      /* distance criterion: removing too short 
         inter-atom distance */
      rsqij = rxij*rxij + ryij*ryij + rzij*rzij;
      if ( rsqij < 1.0 ) {
        ioffset=ioffset+1;
	iok=1;
        break;
      }
    }

    /* update if iok=0 (and otherwise ... ) */
    if ( iok == 0 ) {
      ADp->rx0[ij] = crx0;
      ADp->ry0[ij] = cry0;
      ADp->rz0[ij] = crz0;
      ii=ii+1;
    } else {
      /* if no remaining random variable */
      if ( ird >= nn-3 ) {
        get_rnd_array(&nn,rd);
	ioffset = 1;
	ird = 0;
	icount = icount + 1;
      }
      iok = 0;
    } 
  
    /* check MAXCOUNT */ 
    if ( icount == MAXCOUNT ) {
      ierr = -1;
      break; 
    } 
  }

  /* finalization */
  if ( ierr < 0 ) {
    printf(
      "Initialization of particle position is failed.\n"
      "This error may often consider a too-dense case.\n"
      "You carefully tune total number of particles and\n"
      "unit-cell size.\n");
     return ierr;
  }

  return ierr;
}
#endif
